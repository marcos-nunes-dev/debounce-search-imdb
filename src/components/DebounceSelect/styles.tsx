import styled from 'styled-components';
import { Box } from 'components/Box';
import { theme } from 'styled-tools';

export const SelectWrapper = styled(Box)`
    width: 100%;
    .ant-select-selector,
    .ant-select {
        width: inherit;
    }
    .ant-select-show-search.ant-select-single:not(.ant-select-customize-input)
        .ant-select-selector {
        background: ${theme('colors.gray.50')};
        border: none;
        border-radius: 50px;
    }
    .ant-select-selection-placeholder {
        opacity: 1;
        color: ${theme('colors.gray.300')} !important;
        text-align: center;
    }
`;
