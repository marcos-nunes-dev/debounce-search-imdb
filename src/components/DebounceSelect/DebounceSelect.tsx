import React, { useMemo, useRef } from 'react';
import { SelectWrapper } from './styles';
import { Select as AntdSelect } from 'antd';
import { SelectProps } from 'antd/es/select';
import debounce from 'lodash/debounce';
import { IActor } from 'types/definitions';

export interface DebounceSelectProps<ValueType = any>
    extends Omit<SelectProps<ValueType>, 'options' | 'children'> {
    fetchOptions: (search: string) => Promise<ValueType[]>;
    debounceTimeout?: number;
    options?: ValueType[];
    setOptions: React.Dispatch<React.SetStateAction<any>>;
    onChange: (newValue: unknown, arr: unknown) => void;
}

export function DebounceSelect<
    ValueType extends {
        key?: string;
        label: React.ReactNode;
        value: string | number;
    } = any
>({
    fetchOptions,
    debounceTimeout = 800,
    options,
    setOptions,
    onChange,
    ...props
}: DebounceSelectProps) {
    const fetchRef = useRef(0);

    const debounceFetcher = useMemo(() => {
        const loadOptions = (value: string) => {
            fetchRef.current += 1;
            const fetchId = fetchRef.current;

            value &&
                fetchOptions(value).then((newOptions: IActor[]) => {
                    if (fetchId !== fetchRef.current) return;
                    setOptions(
                        newOptions.slice(0, 10).map((actor: IActor) => ({
                            label: actor.name,
                            value: actor.id,
                            raw: actor,
                        }))
                    );
                });
        };
        return debounce(loadOptions, debounceTimeout);
    }, [fetchOptions, debounceTimeout]);

    return (
        <SelectWrapper>
            <AntdSelect<ValueType>
                filterOption={false}
                onSearch={debounceFetcher}
                notFoundContent={null}
                showArrow={false}
                options={options}
                // Documentation error on listHeight type
                // @ts-ignore
                listHeight='100%'
                onChange={onChange}
                data-testid={'antd-select'}
                {...props}
            />
        </SelectWrapper>
    );
}
