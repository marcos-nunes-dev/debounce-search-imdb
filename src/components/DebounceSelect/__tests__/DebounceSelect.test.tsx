/**
 * @jest-environment jsdom
 */

import React from 'react';
import { render, fireEvent, queryByAttribute, screen } from 'tests';
import { DebounceSelect } from '../DebounceSelect';

const optionsMockup = [{ label: 'hello world', value: '1' }];

describe('<DebounceSelect />', () => {
    it('should render DebounceSelect', () => {
        const setOptMockup = jest.fn();
        const fetchMockup = jest.fn();
        const onChangeMockup = jest.fn();

        const { getByText } = render(
            <DebounceSelect
                options={optionsMockup}
                setOptions={setOptMockup}
                fetchOptions={fetchMockup}
                onChange={onChangeMockup}
                placeholder="ENTER ACTOR'S NAME..."
                debounceTimeout={800}
                showSearch
                labelInValue
                size='large'
            />
        );

        expect(getByText("ENTER ACTOR'S NAME...")).toBeTruthy();
    });

    it('should open default options', () => {
        const setOptMockup = jest.fn();
        const fetchMockup = jest.fn();
        const onChangeMockup = jest.fn();

        const { getByText } = render(
            <DebounceSelect
                options={optionsMockup}
                setOptions={setOptMockup}
                fetchOptions={fetchMockup}
                onChange={onChangeMockup}
                placeholder="ENTER ACTOR'S NAME..."
                debounceTimeout={800}
                showSearch
                labelInValue
                size='large'
            />
        );

        fireEvent.mouseDown(screen.getByTestId('antd-select'));
        setInterval(() => expect(getByText('hello world')).toBeTruthy(), 500);
    });

    it('should fire events', () => {
        const setOptMockup = jest.fn();
        const fetchMockup = jest.fn();
        const onChangeMockup = jest.fn();

        const { getByText } = render(
            <DebounceSelect
                options={optionsMockup}
                setOptions={setOptMockup}
                fetchOptions={fetchMockup}
                onChange={onChangeMockup}
                placeholder="ENTER ACTOR'S NAME..."
                debounceTimeout={800}
                showSearch
                labelInValue
                size='large'
            />
        );

        fireEvent.mouseDown(screen.getByTestId('antd-select'));
        fireEvent.keyDown(screen.getByTestId('antd-select'), {
            key: 'A',
            code: 'KeyA',
        });
        setInterval(() => expect(fetchMockup).toBeCalled(), 900);
        setInterval(() => expect(setOptMockup).toBeCalled(), 900);
    });

    it('should match snapshot', () => {
        const setOptMockup = jest.fn();
        const fetchMockup = jest.fn();
        const onChangeMockup = jest.fn();

        const { container } = render(
            <DebounceSelect
                options={optionsMockup}
                setOptions={setOptMockup}
                fetchOptions={fetchMockup}
                onChange={onChangeMockup}
                placeholder="ENTER ACTOR'S NAME..."
                debounceTimeout={800}
                showSearch
                labelInValue
                size='large'
            />
        );

        expect(container).toMatchSnapshot();
    });
});
