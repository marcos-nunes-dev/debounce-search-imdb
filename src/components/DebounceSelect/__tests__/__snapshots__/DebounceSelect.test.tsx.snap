// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DebounceSelect /> should match snapshot 1`] = `
<div>
  <div
    class="Box-sc-1o1tflv-0 styles__SelectWrapper-sc-fuj6na-0  SXSqq"
  >
    <div
      class="ant-select ant-select-lg ant-select-single ant-select-show-search"
      data-testid="antd-select"
    >
      <div
        class="ant-select-selector"
      >
        <span
          class="ant-select-selection-search"
        >
          <input
            aria-activedescendant="rc_select_TEST_OR_SSR_list_0"
            aria-autocomplete="list"
            aria-controls="rc_select_TEST_OR_SSR_list"
            aria-haspopup="listbox"
            aria-owns="rc_select_TEST_OR_SSR_list"
            autocomplete="off"
            class="ant-select-selection-search-input"
            id="rc_select_TEST_OR_SSR"
            readonly=""
            role="combobox"
            style="opacity: 0;"
            unselectable="on"
            value=""
          />
        </span>
        <span
          class="ant-select-selection-placeholder"
        >
          ENTER ACTOR'S NAME...
        </span>
      </div>
    </div>
  </div>
</div>
`;
